### PR TITLE
Remove the `_default_` mapping.

### PR DESCRIPTION
**Pull request by jpountz**
With only one type, the default mapping has become useless. The approach in
this commit is to reject mapping updates that try to update the `_default`
type regardless of the index creation version, and to ignore `_default_`
mappings when recovering indices from a 6.x index. In the corner case that
an empty 6.x index needs to be upgraded and this index only has a
`_default_` mapping, then the `_default_` mapping will be renamed to `_doc`,
which is the hardcoded mapping name in 7.0.

I need reviews especially on the cluster-state processing side of the PR to know
whether the approach I took in order to dynamically remove default mappings
when processing mappings is ok.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java [hunk 1]
```diff
@@ -255,32 +255,26 @@ private ClusterState applyRequest(ClusterState currentState, PutMappingClusterSt
...
-                                        mapperService.getParentTypes().contains(parentType) == false) {
-                                    throw new IllegalArgumentException("can't add a _parent field that points to an " +
-                                        "already existing type, that isn't already a parent");
-                                }
+                    // TODO: can we find a better place for this validation?
+                    // The reason this validation is here is that the mapper service doesn't learn about
+                    // new types all at once , which can create a false error.
+
+                    // For example in MapperService we can't distinguish between a create index api call
+                    // and a put mapping api call, so we don't which type did exist before.
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java]: nit: s/we don't/we don't know/

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by ywelsch** [comment]
I find the approach chosen here to patch the cluster state problematic, as it leads to inconsistent views across 6.x / 7.x nodes. My suggestion is therefore to not fully remove _default_ in 7.x just yet:
- Keep _default_ in the cluster state on 7.x (but add assertion that it only exists for 6.x indices)
- Reject adding _default_ mappings for 7.x indices (only single mapping in 7.x indices)
- Keep MapperService etc. aware of _default_ mappings because of 6.x indices
- Remove everything in 8.x

**Comment by jpountz** [comment]
Superseded by #29165 where I applied @ywelsch's suggestions.

**Pull request closed without merging**